pkg/operator/controller/route-metrics/controller_test.go

```diff
@@ -1,12 +1,223 @@
+				},
+			},
+			namespace:            unit.NewNamespaceBuilder().WithName("foo-ns").Build(),
+			expectedMetricFormat: routePerShardMetric("foo-ic", 0),
+		},
+		{
+			name:              "ingress controller with no routeSelector and with namespaceSelector with route with no labels in namespace with incorrect labels",
+			ingressController: unit.NewIngressControllerBuilder().WithName("foo-ic").WithAdmitted(true).WithNamespaceSelector("type", "shard").Build(),
+			request:           reconcileRequest,
+			routes: routev1.RouteList{
+				Items: []routev1.Route{
+					*unit.NewRouteBuilder().WithName("foo-route").WithNamespace("foo-ns").WithAdmittedICs("foo-ic").Build(),
+				},
+			},
+			namespace:            unit.NewNamespaceBuilder().WithName("foo-ns").WithLabel("type", "not-shard").Build(),
+			expectedMetricFormat: routePerShardMetric("foo-ic", 0),
+		},
+		{
+			name:              "ingress controller with routeSelector and with namespaceSelector with route with correct labels in namespace with incorrect labels",
+			ingressController: unit.NewIngressControllerBuilder().WithName("foo-ic").WithAdmitted(true).WithRouteSelector("type", "shard").WithNamespaceSelector("type", "shard").Build(),
+			request:           reconcileRequest,
+			routes: routev1.RouteList{
+				Items: []routev1.Route{
+					*unit.NewRouteBuilder().WithName("foo-route").WithNamespace("foo-ns").WithAdmittedICs("foo-ic").WithLabel("type", "shard").Build(),
+				},
+			},
+			namespace:            unit.NewNamespaceBuilder().WithName("foo-ns").WithLabel("type", "not-shard").Build(),
+			expectedMetricFormat: routePerShardMetric("foo-ic", 0),
+		},
+		{
+			name:              "ingress controller with expression routeSelector and with expression namespaceSelector with route with correct labels in namespace with correct labels",
+			ingressController: unit.NewIngressControllerBuilder().WithName("foo-ic").WithAdmitted(true).WithRouteExpressionSelector("type", metav1.LabelSelectorOpIn, []string{"shard"}).WithNamespaceExpressionSelector("type", metav1.LabelSelectorOpIn, []string{"shard"}).Build(),
+			request:           reconcileRequest,
+			routes: routev1.RouteList{
+				Items: []routev1.Route{
+					*unit.NewRouteBuilder().WithName("foo-route").WithNamespace("foo-ns").WithAdmittedICs("foo-ic").WithLabel("type", "shard").Build(),
+				},
+			},
+			namespace:            unit.NewNamespaceBuilder().WithName("foo-ns").WithLabel("type", "shard").Build(),
+			expectedMetricFormat: routePerShardMetric("foo-ic", 1),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var objList []client.Object
+			if tc.ingressController != nil {
+				objList = append(objList, tc.ingressController)
+			}
+			if tc.namespace != nil {
+				objList = append(objList, tc.namespace)
+			}
+			for _, route := range tc.routes.Items {
+				objList = append(objList, route.DeepCopy())
+			}
+			err, _, cache := newFakeClient(objList...)
```

Comment:
To illustrate the idea of the init object list provided for each test case:
```suggestion
		{
			name:              "ingress controller with expression routeSelector and with expression namespaceSelector with route with correct labels in namespace with correct labels",
			initObjs: []client.Object{
			    unit.NewIngressControllerBuilder().WithName("fooic").
			    WithAdmitted(true).
			    WithRouteExpressionSelector("type", metav1.LabelSelectorOpIn[]string{"shard"}).
			    WithNamespaceExpressionSelector("type", metav1.LabelSelectorOpIn, []string{"shard"}).
			    Build(),
			    unit.NewRouteBuilder().
			    WithName("foo-route").
			    WithNamespace("foo-ns").
			    WithAdmittedICs("foo-ic").
			    WithLabel("type", "shard").
			    Build(),
			    unit.NewNamespaceBuilder().
			    WithName("foo-ns").
			    WithLabel("type", "shard").
			    Build(),
			},    
			request:           reconcileRequest,
			expectedMetricFormat: routePerShardMetric("foo-ic", 1),
		},
	}
	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			err, _, cache := newFakeClient(it.initObjs...)
```

Reply:
Oh I see. Basically making the test input "loosely typed". I like it, seems cleaner.

Done